File: langs-vs/Memory.java
public class Memory {

    static class E {
        int a;

        E(int a) {
            this.a = a;
        }
    }

    public static void main(String[] args) {

        final long ARRAY_COUNT = 10000L;
        final long TEST_COUNT  = ARRAY_COUNT * 10_0000L;

        E[] es = new E[(int) ARRAY_COUNT];
        for (long i = 0; i < TEST_COUNT; i++)
             es[(int) (i * 123456789L % ARRAY_COUNT)] = new E((int) i);

        long n = 0;
        for (long i = 0; i < ARRAY_COUNT; i++) {
            E e = es[(int) i];
            if (e != null)
                n += e.a;
        }
        System.out.println(n);
    }
}

File: langs-vs/memory.cpp
#include <memory.h>
#include <stdio.h>

struct E
{
    int a;
    E(int aa) : a(aa) {}
};

int main()
{
    static const long long ARRAY_COUNT = 10000L;
    static const long long TEST_COUNT = ARRAY_COUNT * 100000L;

    E **es = new E *[ARRAY_COUNT];
    memset(es, 0, sizeof(E *) * ARRAY_COUNT);
    for (long long i = 0; i < TEST_COUNT; i++)
    {
        long long idx = i * 123456789L % ARRAY_COUNT;
        E *e = es[idx];
        if (e)
            delete e;
        es[idx] = new E((int)i);
    }

    long long n = 0;
    for (long long i = 0; i < ARRAY_COUNT; i++)
    {
        E *e = es[i];
        if (e)
            n += e->a;
    }
    printf("%lld\n", n);
    return 0;
}
